src/index.ts

```diff
@@ -77,15 +92,47 @@ export const go = async <T, E extends Error>(
   const fn = typeof predicate === 'function' ? predicate : () => predicate;
   if (!options) return attempt(fn);
 
-  const { retries, timeoutMs } = options;
-  const attempts = retries ? retries + 1 : 1;
-  let lastFailedAttemptResult: GoResultError<E> | null = null;
-  for (let i = 0; i < attempts; i++) {
-    const goRes = await attempt<T, E>(fn, timeoutMs);
-    if (goRes.success) return goRes;
+  const { retries, timeoutMs, delay, fullTimeoutMs } = options;
 
-    lastFailedAttemptResult = goRes;
+  let fullTimeoutExceeded = false;
+  let fullTimeoutPromise = new Promise((_resolve) => {}); // Never resolves
+  if (fullTimeoutMs !== undefined) {
+    // Start a "full" timeout that will stop all retries after it is exceeded
+    fullTimeoutPromise = sleep(fullTimeoutMs).then(() => {
+      fullTimeoutExceeded = true;
+      return fail(new Error('Full timeout exceeded'));
+    });
   }
 
-  return lastFailedAttemptResult!;
+  const makeAttempts = async () => {
+    const attempts = retries ? retries + 1 : 1;
+    let lastFailedAttemptResult: GoResultError<E> | null = null;
+    for (let i = 0; i < attempts; i++) {
```

Comment:
Related to comment from other PR: doesn't this cause all requests to be executed all the time? i.e. if `retries = 10`, it will always execute the promise 10 times?

Reply:
We do - https://github.com/api3dao/promise-utils/blob/add-retry-options/src/index.test.ts#L517

Reply:
`for` loops and async JS are complicated 😅 . Wouldn't `for..of` be a better choice here though just to be safe, since that definitely waits?

Reply:
:D Both of them do wait, I don't think its necssary.